plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php

```diff
@@ -0,0 +1,428 @@
+<?php
+declare(strict_types=1);
+
+namespace Automattic\WooCommerce\Blocks\BlockTypes\ProductCollection;
+
+use Automattic\WooCommerce\Blocks\BlockTypes\AbstractBlock;
+use InvalidArgumentException;
+use WP_Query;
+use WC_Tax;
+
+/**
+ * ProductCollectionController class.
+ */
+class ProductCollectionController extends AbstractBlock {
+
+	/**
+	 * Block name.
+	 *
+	 * @var string
+	 */
+	protected $block_name = 'product-collection';
+
+	/**
+	 * Instance of CollectionHandlerRegistry.
+	 *
+	 * @var CollectionHandlerRegistry
+	 */
+	protected $collection_handler_registry;
+
+	/**
+	 * The render state of the product collection block.
+	 *
+	 * These props are runtime-based and reinitialize for every block on a page.
+	 *
+	 * @var array
+	 */
+	private $render_state = array(
+		'has_results'          => false,
+		'has_no_results_block' => false,
+	);
+
+	/**
+	 * Instance of ProductQueryBuilder.
+	 *
+	 * @var ProductQueryBuilder
+	 */
+	protected $query_builder;
+
+	/**
+	 * Instance of ProductRenderer.
+	 *
+	 * @var ProductRenderer
+	 */
+	protected $renderer;
+
+	/**
+	 * Initialize this block type.
+	 *
+	 * - Hook into WP lifecycle.
+	 * - Register the block with WordPress.
+	 * - Hook into pre_render_block to update the query.
+	 */
```

Comment:
Nitpick: This comment looks outdated.

Reply:
Fixed in commit [06acc33](https://github.com/woocommerce/woocommerce/pull/52429/commits/06acc3331538ee176d2ccf95846df85335840427)!